### PR TITLE
[AI Chat] Create Chat Workspace

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR adds the `ChatWorkspace`, `ChatMessage`, and `UserChatMessageEditor` .tsx components and displays a very basic version of the AI chat workspace (without the Instructions panel). I added the styling via 'module.scss' files. I used `MusicView.jsx` as a reference as well as [Ben's PR](https://github.com/code-dot-org/code-dot-org/pull/41478) that created the first iteration of the Java Lab Code Review. Note that I added a few demo messages in `demoMessages.ts` temporarily.

Screenshot with only 3 visible chat messages displayed:

Screen Shot 2023-08-03 at 9 23 29 AM

Screen Shot 2023-08-03 at 9 22 36 AM

Screenshot with 5 messages including user messages that are designated as either too personal or inappropriate:

Screencast video of 10 messages so that scrolling is required:

https://github.com/code-dot-org/code-dot-org/assets/107423305/1d19d35e-9128-4580-ad03-ff8923ce6b87

~~Note that the user chat message editor is not positioned at the bottom of the screen. I would love help with positioning it at the bottom of the container.~~ This is now resolved.

## Links
[jira](https://codedotorg.atlassian.net/browse/SL-1047)
[Figma](https://www.figma.com/file/nyvdBdwpdKSkOR082LVWQc/ClassGPT?type=design&node-id=211-2&mode=design&t=U16ZutHVsbPIgOL1-0)

## Testing story
Testing locally.

## Follow-up work
Add auto-scrolling in chat conversation area - based on [this comment](https://github.com/code-dot-org/code-dot-org/pull/53080#issuecomment-1659617170).
[jira](https://codedotorg.atlassian.net/browse/SL-1049)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/aichat/views/ChatMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  } else if (status === 'personal') {
+    return (
+      <div
+        id={'chat-workspace-message-body-too-personal'}
+        className={classNames(styles.message, styles.tooPersonalMessage)}
+      >
+        {TOO_PERSONAL_MESSAGE}
+      </div>
+    );
+  }
```
Comment (fisher-alice) [inline apps/src/aichat/views/ChatMessage.tsx:56]: Unsure if I need an `else` to return a `<div></div>`.

File: apps/src/aichat/views/chatMessage.module.scss [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+.message {
+    background-color: $lightest_gray;
+    padding: 10px 12px;
+}
+
+.userMessageContainer {
+    margin: 0 0 25px 50px;
+    border-radius: 10px;
```
Comment (fisher-alice) [inline apps/src/aichat/views/chatMessage.module.scss]: Not sure why `border-radius` here and below are not being set by this stying. Any advice is appreciated~

File: apps/src/aichat/views/ChatMessage.tsx (hunk 1 above)
Comment (breville) [reply apps/src/aichat/views/ChatMessage.tsx:56]: It [seems](https://react.dev/learn/conditional-rendering#conditionally-returning-nothing-with-null) like `null` is appropriate?

File: apps/src/aichat/views/chatWorkspace.module.scss [hunk 3]
```diff
@@ -0,0 +1,4 @@
+.chatArea {
+    margin: 10px 5%;
+    overflow: 'auto';
```
Comment (breville) [inline apps/src/aichat/views/chatWorkspace.module.scss]: None of these should have quotes since they are real CSS rather than JS.

File: apps/src/aichat/views/ChatMessage.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+const isUser = (role: string) => {
+  return role === 'user';
+};
+
+const displayUserMessage = (status: string, chatMessageText: string) => {
+  if (status === 'ok') {
+    return (
+      <div
+        id={'chat-workspace-message-body'}
```
Comment (breville) [inline apps/src/aichat/views/ChatMessage.tsx]: These IDs are handy for debugging, but for elements that are used more than once, it should be a class name rather than an ID, because each ID should only appear in the DOM once.

File: apps/src/aichat/views/chatMessage.module.scss (hunk 2 above)
Comment (breville) [reply apps/src/aichat/views/chatMessage.module.scss]: It's the child that has the background colour, so this radius will only do something if the overflow is hidden, but the quotes around the hidden are preventing it from working.  

That said, because this parent is larger than the child, you're not really getting the rounding exactly where you want it.  It seems more appropriate to move the border radius to the child (currently known as `#chat-workspace-message-body`) and then no hidden overflow is needed.

File: apps/src/aichat/views/AichatView.tsx [hunk 5]
```diff
@@ -3,38 +3,23 @@
...
-        >
-          {aichatLocale.continue()}
-        </button>
-      </Aichat>
+    <div id="aichat" className={moduleStyles.aichat}>
+      {/* <div id="chat-instructions-area" className={moduleStyles.instructions}>
+        <PanelContainer id="chat-instructions-panel" headerText="Instructions">
+          <Instructions />
+        </PanelContainer>
+      </div> */}
```
Comment (breville) [inline apps/src/aichat/views/AichatView.tsx]: If code is commented out, there should be a comment explaining why, but it's probably better to remove it and then bring it back later if it becomes necessary.

**Comment by breville** [comment]
> Note that the user chat message editor is not positioned at the bottom of the screen. I would love help with positioning it at the bottom of the container.

You can take a look at how https://studio.code.org/projectbeats uses flexbox to pin the fixed-height bottom part (`#play-area`) to the bottom, while the top part (`#work-area`) dynamically fills the remainder.

You'll probably also want some kind of auto-scrolling strategy so that new chat entries force the upper area to scroll to its bottom, at least if the user was already scrolled to the existing bottom, while the user can still manually scroll higher.  Maybe you can show a little indicator if new text entries appear while they aren't scrolled to the bottom, and maybe clicking on it can even scroll smoothly down to the bottom.  `scrollIntoView` [looks](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) promising.

File: apps/src/aichat/views/chatMessage.module.scss (hunk 2 above)
Comment (fisher-alice) [reply apps/src/aichat/views/chatMessage.module.scss]: Thanks so much! This makes a lot of sense. Moved the border radius styling to `.message`.
I also removed the user name according to the Figma design.

Screen Shot 2023-08-01 at 9 31 27 AM

File: apps/src/aichat/views/ChatMessage.tsx (hunk 4 above)
Comment (fisher-alice) [reply apps/src/aichat/views/ChatMessage.tsx]: Thanks! I'll go ahead and remove the `id` since I have the classNames included by type of message (user, inappropriate, or tooPersonal).

File: apps/src/aichat/views/AichatView.tsx (hunk 5 above)
Comment (fisher-alice) [reply apps/src/aichat/views/AichatView.tsx]: Will remove for now.

**Comment by fisher-alice** [comment]
> looks

Thanks Brendan - I placed the message editor at the bottom this [this commit](https://github.com/code-dot-org/code-dot-org/pull/53080/commits/9552102ef5225eba0523cef9ea1a1c12b7ff6594). Then I started using `flexbox` in the module file but getting some interesting and undesirable layout results. Looking forward to chatting about this in our 1-1 today - thanks! 


Screen Shot 2023-08-01 at 11 08 17 AM

File: apps/src/aichat/views/AichatView.tsx [hunk 6]
```diff
@@ -3,38 +3,46 @@
...
-          type="button"
-          onClick={() => nextButtonPressed()}
-          className={styles.buttonNext}
-        >
-          {aichatLocale.continue()}
-        </button>
-      </Aichat>
+    <div id="aichat-lab" className={moduleStyles.aichatLab}>
+      <div id="chat-workspace-area">
+        <PanelContainer id="chat-workspace-panel" headerText="AI Chat">
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/AichatView.tsx]: Can we use a localized string here (for "AI Chat") / start a localization file if it doesn't exist yet? Or this is meant to be a temporary label?

File: apps/src/aichat/views/AichatView.tsx [hunk 7]
```diff
@@ -3,38 +3,46 @@
...
-          id="aichat-continue-button"
-          type="button"
-          onClick={() => nextButtonPressed()}
-          className={styles.buttonNext}
-        >
-          {aichatLocale.continue()}
-        </button>
-      </Aichat>
+    <div id="aichat-lab" className={moduleStyles.aichatLab}>
+      <div id="chat-workspace-area">
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/AichatView.tsx]: Might be good to pull out the "chat-workspace-area" into its own component as well, just to make sure this top level component doesn't get too bloated? that's something we ran into with Music Lab where in the beginning, we added most functionality directly to our top-level view (MusicView). this made the file pretty hard to parse as we added more and more, until we started intentionally removing code from it

File: apps/src/aichat/views/ChatMessage.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+interface ChatMessageProps {
+  id: string;
+  name: string;
+  role: string;
+  chatMessageText: string;
+  status: string;
+}
+
+const INAPPROPRIATE_MESSAGE = 'This message has been flagged as inappropriate.';
+const TOO_PERSONAL_MESSAGE = 'This message has been flagged as too personal.';
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatMessage.tsx]: Localization for these as well? ok if this is being handled as a follow up

File: apps/src/aichat/views/ChatMessage.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+import styles from './chatMessage.module.scss';
+import classNames from 'classnames';
+
+interface ChatMessageProps {
+  id: string;
+  name: string;
+  role: string;
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatMessage.tsx]: If we know that `role` can only be `"assistant"` or `"student"`, we can just make this type `role: 'student' | 'assistant'` to guarantee it's one of the two.

File: apps/src/aichat/views/ChatMessage.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  chatMessageText: string;
+  status: string;
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatMessage.tsx]: Same for `status` as well (`status: 'ok' | 'inappropriate' | 'personal'`)

File: apps/src/aichat/views/UserChatMessageEditor.tsx [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        placeholder="Add a chat message..."
+      />
+      <Button
+        key="submit"
+        text="Submit"
+        icon="arrow-up"
+        onClick={() => {
+          console.log('Submit button clicked');
+        }}
+        color="purple"
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/UserChatMessageEditor.tsx:19]: Can this be set in CSS?

File: apps/src/aichat/views/aichat.module.scss [hunk 12]
```diff
@@ -13,20 +52,8 @@
...
-  width: 100%;
-  height: 100%;
-  top: 0;
-  left: 0;
-  flex-direction: column;
-}
-
-.buttonNext {
-  animation: fade-in-button 7s;
+.nextButton {
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/aichat.module.scss]: Do we have a nextButton? Can we get rid of this?

File: apps/src/aichat/views/chatMessage.module.scss [hunk 13]
```diff
@@ -0,0 +1,39 @@
+@import 'color.scss';
+
+.name {
+    font-family: "Gotham 5r", sans-serif;
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/chatMessage.module.scss]: Probably a question for @levadadenys - should we use a `<Typography>` component in place of this?

File: apps/src/aichat/views/chatMessageShape.ts [hunk 14]
```diff
@@ -0,0 +1,91 @@
+import PropTypes from 'prop-types';
+
+const chatMessageShape = PropTypes.shape({
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/chatMessageShape.ts]: Is this used?

**Comment by fisher-alice** [comment]
> > looks
> 
> Thanks Brendan - I placed the message editor at the bottom this [this commit](https://github.com/code-dot-org/code-dot-org/pull/53080/commits/9552102ef5225eba0523cef9ea1a1c12b7ff6594). Then I started using `flexbox` in the module file but getting some interesting and undesirable layout results. Looking forward to chatting about this in our 1-1 today - thanks!

Thanks for your help, @breville ! I added `display: flex`, `overflow: scroll`, and `flex-direction: column` to `chatWorkspace` and then set `conversationArea`'s `flex-grow: 1` and the editor's height to a fixed value. I'll add an auto-scrolling strategy in another PR as a follow-up.

File: apps/src/aichat/views/AichatView.tsx (hunk 7 above)
Comment (fisher-alice) [reply apps/src/aichat/views/AichatView.tsx]: Sure! I actually had a `chatWorkspace` in this [earlier commit](https://github.com/code-dot-org/code-dot-org/pull/53080/commits/18b0f5c4417e2b93fb7348fa302c10cd7e751bd8) but was trying to fix the CSS `flex` display so will add it back now - thanks!

File: apps/src/aichat/views/AichatView.tsx (hunk 6 above)
Comment (fisher-alice) [reply apps/src/aichat/views/AichatView.tsx]: Will do - thanks!

File: apps/src/aichat/views/ChatMessage.tsx (hunk 8 above)
Comment (fisher-alice) [reply apps/src/aichat/views/ChatMessage.tsx]: I'll go ahead and add in this PR.

File: apps/src/aichat/views/ChatMessage.tsx (hunk 9 above)
Comment (fisher-alice) [reply apps/src/aichat/views/ChatMessage.tsx]: Cool - will do!

File: apps/src/aichat/views/chatMessage.module.scss (hunk 13 above)
Comment (fisher-alice) [reply apps/src/aichat/views/chatMessage.module.scss]: I'll defer to @levadadenys . Thanks!

File: apps/src/aichat/views/chatMessageShape.ts (hunk 14 above)
Comment (fisher-alice) [reply apps/src/aichat/views/chatMessageShape.ts]: ~~@bencodeorg said it might be useful for him as he works on front-end client so i'll keep in place temporarily.~~
Since the props are defined clearly in `ChatMessage.tsx`, I'll go ahead and remove. Thanks!

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 11 above)
Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx:19]: I tried adding to `userChatMessageEditor.module.scss` the following:
```
@import 'color.scss';

.Button {
    color: $purple;
}
```
But the style didn't take. Do you have any suggestions? Thanks!

Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx:19]: I'm importing `Button` from 'templates/Button.jsx'.

File: apps/src/aichat/views/demoMessages.ts [hunk 15]
```diff
@@ -0,0 +1,81 @@
+// Will delete once we've connected UI to backend.
+const demoChatMessages = [
+  {
+    id: '1',
+    name: 'HistoryBot',
+    role: 'assistant' as const,
```
Comment (fisher-alice) [inline apps/src/aichat/views/demoMessages.ts]: @sanchitmalhotra126 - I received an error unless I assigned `role` as above (with `as const`) after I defined the `role` types as either `system`, `user`, or `assistant`. @bencodeorg - heads up.

File: apps/src/aichat/views/AichatView.tsx (hunk 7 above)
Comment (fisher-alice) [reply apps/src/aichat/views/AichatView.tsx]: Hey Sanchit - I added `chatWorkspace` back, but then 'flex' doesn't seem to be working as expected. I posted a couple screencast videos in the updated description above so removing `chatWorkspace` from `AichatView` for now.

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 11 above)
Comment (sanchitmalhotra126) [reply apps/src/aichat/views/UserChatMessageEditor.tsx:19]: I'm guessing you'll have to provide a `className` to the button component for this to work

File: apps/src/aichat/views/demoMessages.ts [hunk 16]
```diff
@@ -0,0 +1,89 @@
+// Will delete once we've connected UI to backend.
+
+const ASSISTANT = 'assistant' as const;
+const USER = 'user' as const;
+const SYSTEM = 'system' as const;
+const OK = 'ok' as const;
+const PERSONAL = 'personal' as const;
+const INAPPROPRIATE = 'inappropriate' as const;
+
+const demoChatMessages = [
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/demoMessages.ts]: Hm, I wonder if you declare the type of this array, you may not need to use `as const` on all of those values? Especially if this chat message structure is going to be shared between the front end and the back end, it would probably be valuable to pull it out from the React component and into its own type, stored in something like `src/aichatc/types.ts`. Then you could declare this array as `const demoChatMessages: ChatMessage[] = [ ... ]`.

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 11 above)
Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx:19]: Thanks. I did try to add `className={moduleStyles.submitButton}` to `Button` with the following added to the module file:
```
@import 'color.scss';

.submitButton {
    color: $purple;
}
```
But this didn't seem to work either.

File: apps/src/aichat/views/demoMessages.ts (hunk 16 above)
Comment (fisher-alice) [reply apps/src/aichat/views/demoMessages.ts]: Thanks Sanchit! I haven't added TypeScript types before so will look into this.

Comment (fisher-alice) [reply apps/src/aichat/views/demoMessages.ts]: Just added a `ChatMessage` type and no longer need `as const` in demo messages. Thanks!

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 11 above)
Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx:19]: After discussion offline, will keep `Button` as is.

File: apps/src/aichat/views/AichatView.tsx (hunk 7 above)
Comment (fisher-alice) [reply apps/src/aichat/views/AichatView.tsx]: This is now resolved. I learned that the `flex` styling is not inherited from parent and must include in the div.

File: apps/src/aichat/views/chatMessage.module.scss (hunk 13 above)
Comment (fisher-alice) [reply apps/src/aichat/views/chatMessage.module.scss]: Added `Typography` and will ask @levadadenys to take a look when he gets back so he can tweak as he sees fit.

File: apps/src/aichat/views/UserChatMessageEditor.tsx [hunk 17]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import Button from '@cdo/apps/templates/Button';
+import moduleStyles from './userChatMessageEditor.module.scss';
+
+const UserChatMessageEditor: React.FunctionComponent = () => {
+  return (
+    <div className={moduleStyles.UserChatMessageEditor}>
+      <textarea
+        className={moduleStyles.textArea}
+        placeholder="Add a chat message..."
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/UserChatMessageEditor.tsx:10]: localization for this too?

**Review by sanchitmalhotra126 — APPROVED** [review]
Looks good!

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 17 above)
Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx:10]: Oh yes! I'll add the localized string in my follow-up PR. Thank you so much for all your feedback and help!
